pip/pip-321.md

```diff
@@ -0,0 +1,241 @@
+
+When a namespace has the policy with `allowed_clusters` and `replication_clusters`, the topics under this namespace will replicate data to all `replication_clusters` by default. Additionally, they can have their own flexible replication configuration, which should be a subset of the `allowed_clusters`.
+
+If `allowed_clusters` is not set, `replication_cluster` will be used as the default value for `allowed_cluster`.
+
+If neither `allowed_clusters` nor `replication_cluster` are set, topics under this namespace will only be able to publish/subscribe at the local cluster.
+
+Message-level replication is similar to topic-level replication. The replication clusters of a message should be the subset of the `allowed_cluster`, and are the `replication_clusters` configured at the topic level or namespace level by default.
+
+# Detailed Design
+
+## Public-facing Changes
+
+### Public API
+
+#### `setNamespaceAllowedClusters` Endpoint
+
+This new endpoint allows setting the list of allowed clusters for a specific namespace.
+
+**Method:**
+```
+POST
+```
+
+**Path:**
+```
+/namespaces/{tenant}/{namespace}/allowedclusters
+```
+
+**Query Parameters:**
+
+- `tenant`: The tenant within which the namespace resides.
+- `namespace`: The namespace for which you are setting the allowed clusters.
+
+**HTTP Body Parameters:**
+
+- `clusterIds`: A list of cluster IDs.
+
+**Response Codes:**
+
+- `400 Bad Request`: The list of allowed clusters should include all replication clusters.
+- `403 Forbidden`: The requester does not have admin permissions.
+- `404 Not Found`: The specified tenant, cluster, or namespace does not exist.
+- `409 Conflict`: A peer-cluster cannot be part of an allowed-cluster.
+- `412 Precondition Failed`: The namespace is not global or the provided cluster IDs are invalid.
+
+**Explanation for 409 Conflict:** This follows the behavior of namespace replication clusters. As per PIP-8, a peer-cluster cannot be part of a replication-cluster. Similarly, for allowed-clusters, users could enable replication at the topic level, hence a peer-cluster cannot be part of allowed-clusters as well.
+
+#### `getNamespaceAllowedClusters` Endpoint
+
+This new endpoint allows retrieving the list of allowed clusters for a specific namespace.
+
+**Method:**
+```
+GET
+```
+
+**Path:**
+```
+/namespaces/{tenant}/{namespace}/allowedclusters
+```
+
+**Query Parameters:**
+
+- `tenant`: The tenant within which the namespace resides.
+- `namespace`: The namespace for which you are retrieving the allowed clusters.
+
+**Response Codes:**
+
+- `403 Forbidden`: The requester does not have admin permissions.
+- `404 Not Found`: The specified tenant, cluster, or namespace does not exist.
+- `412 Precondition Failed`: The namespace is not global.
+
+### Binary protocol
+
+### Configuration
+
+### CLI
```

Comment:
Please also add the CLI changes to `bin/pulsar-admin namespaces create`

```
bin/pulsar-admin namespaces create
Main parameters are required ("tenant/namespace")

Creates a new namespace
Usage: create [options] tenant/namespace
  Options:
    --bundles, -b
      number of bundles to activate
      Default: 0
    --clusters, -c
      List of clusters this namespace will be assigned
```

To allow users to set allowed clusters when creating a namespace

Reply:
I am wondering whether we can mark the following Admin API as `@Deprecated`
```
void createNamespace(String namespace, Set<String> clusters) throws PulsarAdminException;
```
Because the Admin API with `Policies` is enough.
```
void createNamespace(String namespace, Policies policies) throws PulsarAdminException;
```

Reply:
It is used to set `clusterID` for v1 admin.
For v2 admin, creating namespaces with policy is enough.
```java
        if (ns.isV2()) {
            // For V2 API we pass full Policy class instance
            Policies policies = new Policies();
            policies.replication_clusters = clusters;
            return asyncPutRequest(path, Entity.entity(policies, MediaType.APPLICATION_JSON));
        } else {
            // For V1 API, we pass the BundlesData on creation
            return asyncPutRequest(path, Entity.entity("", MediaType.APPLICATION_JSON)).thenCompose(ignore -> {
                // For V1, we need to do it in 2 steps
                return setNamespaceReplicationClustersAsync(namespace, clusters);
            });
        }
```

Reply:
Oh, I got your point.
It should be fine. `void createNamespace(String namespace, Set<String> clusters) throws PulsarAdminException;` is more intuitive for users to create namespaces with replication clusters. Policies usually means TTL, data retention ...

Reply:
The `replication_clusters ` also is a namespace policy. And after this proposal, do we need to add `createNamespace(String namespace, Set<String> allowedClusters)` and `createNamespace(String namespace, Set<String> clusters, Set<String> allowedClusters)`?  
I think this would add two unnecessary methods. Moreover, the parameter name `clusters` in the method parameters could confuse users.

Reply:
In my view, this method is merely used for maintaining compatibility with admin1, hence its method parameters have limitations. We should mark it as deprecated.
`void createNamespace(String namespace, Set<String> clusters) throws PulsarAdminException;`